### DISCUSSION TITLE
Request for Granular Folder-Level Access Control in Github

### DISCUSSION THREAD

**Discussion by rakibulalam** [community/community#102755] [Code Security]
### Select Topic Area

General

### Body

I am writing to highlight a significant challenge faced by developers working on projects that use a **mono-repository framework** to manage microservices, frontend and backend applications. The current lack of granular folder-level access control within these mono repositories creates security and access management issues for development teams.

### **Problem Statement:**
With the growing trend towards **mono repositories** for managing microservices, frontend and backend applications, developers often find themselves working on projects with a unified codebase. While this approach has many advantages, it also presents challenges related to access control. Currently, developers with access to the Mono repository gain visibility and access to all folders, including microservices, frontend, and backend applications.

### **Problems encountered:**

- **Security concerns:** Housing of different services in mono repositories, lack of fine-grained access control at the folder level poses risk of unintended access to sensitive code and data.

- **Access Management:** With multiple teams responsible for different services within a mono repository, managing access permissions becomes complex. Developers may have wider access than necessary, leading to potential security vulnerabilities.

- **Codebase clutter:** In large mono repositories, the presence of multiple services can lead to codebase clutter. Developers may inadvertently interact with or modify code outside of their designated area.

**Suggested solution:**
I urge GitHub to consider implementing a feature that allows for granular folder-level access control within Mono repositories. This enhancement will enable developers to set permissions at the folder level, ensuring that each team or developer has access to specific parts of the codebase related to their responsibilities.

### **Use Cases:**

- **Microservices:** Developers working on a specific microservice should only have access to folders and code related to that specific microservice.

- **Separation of frontend and backend:** Teams responsible for frontend and backend applications may have separate access permissions to maintain separation of concerns.

- **Enhanced security:** Folder-level access control supports a more secure development environment, reducing the risk of unintended access to sensitive information.

I believe that including folder-level access control in **Mono repositories** will significantly increase the manageability and security of projects following this architecture. I appreciate GitHub's commitment to continuous improvement and innovation, and I look forward to any updates on this.

**Comment by runephilosof-abtion** [comment]
You can limit write access by setting branch protection rules.
But you cannot limit read access. I don't think that is something Github can implement, that is part of how git works.
You will have to split your codebase up in different repos, possibly connecting them in a mono repo as git modules.

**Reply by rakibulalam** [reply]
@runephilosof-abtion  With the increasing prevalence of mono repo and microservices-based applications, safeguarding read access becomes imperative. This is precisely need GitHub is to consider for introducing features to protect read access. Creating different repositories for a unified code system will be hard to maintain.

**Reply by runephilosof-abtion** [reply]
How do you propose the workflow would be?
What should happen when a user run `git clone` on their local machine?

**Reply by rakibulalam** [reply]
@runephilosof-abtion  thanks for your reply in this proposal can be effective:

Workflow:

1. Create a file named `gitaccess` in the root directory to serve as the access path configuration.
2. Update the manage access to include the new access path options.
3. Implement an access path protection controller to effectively manage access rights.
4. Ensure that collaborators can easily view any restrictions specified in the `.gitaccess` file.
5. If any necessary access is missing, collaborators should promptly inform the appropriate person for authorization.
6. Whenever new access is granted by the root, ensure it is promptly documented in the `.gitaccess` file.
7. Make sure that the `.gitaccess` file is correctly permissioned by the root user to maintain security.

**Reply by runephilosof-abtion** [reply]
I meant more low level, how should this work with the git protocol, when the collaborator checks out the repository locally.
Git is a distributed version control system, so I don't see how this can work on any git repository.

**Reply by webdevs0112** [reply]
@rakibulalam, Your request for granular folder-level access control within GitHub mono repositories is well-articulated and addresses a critical need for developers working on projects with a unified codebase. Implementing such a feature would undoubtedly enhance security, access management, and overall codebase organization.

Here are some additional points to consider reinforcing your request:

1. **Compliance Requirements**: Emphasize how granular folder-level access control can help organizations meet compliance requirements such as GDPR, HIPAA, or industry-specific regulations by ensuring that only authorized individuals have access to sensitive data.

2. **Collaboration Efficiency**: Highlight how this feature can streamline collaboration by allowing teams to focus on their respective areas of responsibility without the distraction of unrelated code. This can lead to increased productivity and smoother development workflows.

3. **Customizable Permissions**: Consider suggesting the ability to customize permissions at the folder level, such as read-only access, write access, or restricted access, to provide flexibility in managing team contributions.

4. **Visibility Control**: Discuss the importance of enabling administrators to control the visibility of specific folders to different teams or individuals, ensuring that sensitive or proprietary code remains confidential.

5. **Integration with Existing Access Control Mechanisms**: Propose integrating folder-level access control seamlessly with GitHub's existing access control mechanisms, such as repository permissions and branch protections, to provide a unified and consistent user experience.

By including these additional points and providing specific use cases and potential benefits, you can further strengthen your request and demonstrate the value of implementing granular folder-level access control within GitHub's mono repositories.

I hope this help you.
Thank you.

**Reply by edahlseng** [reply]
@runephilosof-abtion, is this something that [Git's sparse-checkout feature](https://git-scm.com/docs/git-sparse-checkout) could help with? For example, if someone doesn't have read access to every path in a Git repo, GitHub could deny the request unless a sparse-checkout is used that only access approved paths?

As far as what to show on the web GUI, I imagine that's much more straightforward.

**Comment by zoltan-spire** [comment]
My use case would be to create a GitOps workflow for managing configuration files where all users have read access to all folders and files, write access is restricted for each user/team to a certain folder.

I believe this would be possible without getting tangled up in the git protocol, because it would simply be a pre-git level access control check: if the commit has changes anywhere outside the permitted folders the push is rejected.

**Reply by runephilosof-abtion** [reply]
That is already possible with branch protection rules.

**Reply by zoltan-spire** [reply]
From what I understand from [branch protection rules](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/managing-a-branch-protection-rule) the patterns only apply to branch names, not the folders and file paths.

Perhaps combining branch protection with [CODEOWNERS](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) would be an option to achieve something similar? Or is there a simpler way of doing this?
